compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -723,7 +723,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     // Check if async function's return type was omitted.
                     // Don't emit suggestions if the found type is `impl Future<...>`.
                     debug!("suggest_missing_return_type: found = {:?}", found);
-                    if self.get_impl_future_output_ty(found).is_none() {
+                    if found.is_suggestable(self.tcx, true) {
```

Comment:
Using `is_suggestable` with `false` is actually consistent with other diagnostics, which don't suggest anything if there's inference variables. See:

```
fn foo() {
    Ok(())
}
```

which does *not* suggest `-> Result<(), _>`, since that's not exactly complete. 

I'm not sure if we should be suggesting something when there are inference variables in the return type -- you could probably just come up with a different example instead of `Ok(())` in the UI test you added.

Either that, or the regular "return type is missing" suggestion should be modified, but then the subdiagnostic applicability would need to be bumped down to "has-placeholders" from "machine-applicable".

Reply:
Yeah that makes sense, I will change it back and modify the test

Reply:
Thanks! Maybe worth adding a test for something that's "not suggestable" too, like a closure.